### PR TITLE
net: dhcpv4_server: Improve declined addresses management

### PR DESCRIPTION
**Pull request by rlubos**
In case conflict is detected (either due to receiving Decline message or
due to ICMP probe getting reply), the conflicting address becomes
blocked for further use.

Although the RFC is not specific about how long should the address be
blocked, it make sense to implement some fallback mechanisms to reuse
blocked addresses in the server, otherwise, after longer period of
operation, it may run out of usable address.

This commit adds a timeout for declined addresses, so that by default
the address is marked back as "free" after 24 hrs (default lease time).
It also implements a mechanism, which allows to re-use the oldest
declined entry in case the server runs out of fresh addresses to assign.

### CODE REVIEW

File: subsys/net/lib/dhcpv4/Kconfig [hunk 1]
```diff
@@ -85,6 +85,17 @@ config NET_DHCPV4_SERVER_ADDR_LEASE_TIME
 	  The lease time determines when the IPv4 address lease expires if the
 	  client does not renew it.
 
+config NET_DHCPV4_SERVER_ADDR_DECLINE_TIME
+	int "The time IPv4 addresses remains blocked after conflict detection (seconds)"
+	default 86400
+	help
```
Comment (jukkar) [inline subsys/net/lib/dhcpv4/Kconfig:91]: Would it make sense to add a range for allowable values?
We do not have ranges for other int values in this kconfig either which might be a bit problematic especially as we do not check the value in the code.

**Review by jukkar — COMMENTED** [review]
Overall LGTM, minor nit about allowed values of the Kconfig option(s).

File: subsys/net/lib/dhcpv4/Kconfig (hunk 1 above)
Comment (rlubos) [reply subsys/net/lib/dhcpv4/Kconfig:91]: I just wonder what would be correct range values, it's not specified by RFC so we'd need to choose some arbitrary ones. That's why I've left it w/o range.

Comment (jukkar) [reply subsys/net/lib/dhcpv4/Kconfig:91]: I was mostly concerned if user sets timeout as negative value, and that value is fed directly to `K_SECONDS()`, what is the timeout value in that case. This is of course quite academic, why would anyone put <0 value here.
Anyway, I was just thinking what could go wrong. We could limit the value to [0, MAX_INT] for example.

Comment (rlubos) [reply subsys/net/lib/dhcpv4/Kconfig:93]: Good point with the negative values. Ok, I've applied ranges for the time options, for lease (and decline) time uint32_t max value, as it's the theoretical maximum that lease time option can carry. For probe timeout, 60 seconds max, the maximum is already high enough to make the server replies sluggish, I don't think setting anything higher would make sense at all.

**Review by jukkar — APPROVED** [review]
Thanks, LGTM

**Review by krish2718 — APPROVED** [review]

**Pull request merged**